docs/function_guide.rst

```diff
@@ -0,0 +1,299 @@
+==================================
+Writing libEnsemble User Functions
+==================================
+
+libEnsemble coordinates ensembles of calculations performed by three main
+functions: a :ref:`Generator Function<api_gen_f>`, a :ref:`Simulator Function<api_sim_f>`,
+and an :ref:`Allocation Functions<api_alloc_f>`, or ``gen_f``, ``sim_f``, and
+``alloc_f`` respectively. These are all referred to as User Functions. Although
+libEnsemble includes several ready-to-use User Functions like
+:doc:`APOSMM<examples/aposmm>`, it's expected many users will write their own or
+adjust included functions for their own use-cases.
+This guide serves as an overview of both necessary and optional components for
+writing different kinds of User Functions, and common development patterns.
+
+Generator Functions
+===================
+
+As described in the :ref:`API<api_gen_f>`, the ``gen_f`` is called by a
+libEnsemble worker via the following::
+
+    out = gen_f(H[gen_specs['in']][sim_ids_from_allocf], persis_info, gen_specs, libE_info)
+
+In practice, most ``gen_f`` function definitions written by users resemble::
+
+    def my_generator(H, persis_info, gen_specs, libE_info):
+
+Where :doc:`H<data_structures/history_array>` is a selection of the
+:doc:`History array<history_output>`, determined by sim IDs from the
+``alloc_f``, :doc:`persis_info<data_structures/persis_info>` is a dictionary
+containing state information, :doc:`gen_specs<data_structures/gen_specs>` is a
+dictionary containing pre-defined parameters for the ``gen_f``, and ``libE_info``
+is a dictionary containing libEnsemble-specific entries. See the API above for
+more detailed descriptions of the parameters.
+
+.. note::
+
+    If the ``gen_f`` is a persistent generator, then ``gen_specs`` will often be
+    empty if the ``alloc_f`` determines what fields to send to the generator.
+
+Typically users start by parsing their custom parameters initially defined
+within ``gen_specs['user']`` in the calling script and defining a *local* History
+array based on the datatype in ``gen_specs['out']``, to be returned. For example::
+
+        batch_size = gen_specs['user']['batch_size']
+        local_H_out = np.zeros(batch_size, dtype=gen_specs['out'])
+
```

Comment:
It might be a bit advanced here. But it is possible, if you want to update existing fields to provide a sim_id and some subfield and it will just update that field. This is how I cancelled simulations from the gen in cwp that have already been generated.

H_o = np.zeros(len(sim_ids_to_cancel), dtype=[('sim_id', int), ('cancel', bool)])

@jmlarson1 any opinion on if we should include here.

Reply:
Perhaps we should wait for a future version. As this also goes with a more irregular comms pattern.